pliers/extractors/audio.py

```diff
@@ -488,4 +494,140 @@ class PercussiveExtractor(LibrosaFeatureExtractor):
+            verify_dependencies(['yamnet'])
+        except MissingDependencyError: 
+            raise MissingDependencyError(dependencies=None,
+                                         custom_message=YAMNET_INSTALL_MESSAGE)
+        verify_dependencies(['tensorflow'])
+
+        MODULE_PATH = path.dirname(yamnet.__file__)
+        LABELS_PATH = path.join(MODULE_PATH, 'yamnet_class_map.csv')
+        self.weights_path = weights_path or path.join(MODULE_PATH, 'yamnet.h5')
+        self.hop_size = hop_size
+        self.yamnet_kwargs = yamnet_kwargs
+        self.params = yamnet.params
+        self.params.PATCH_HOP_SECONDS = hop_size
+        for par, v in self.yamnet_kwargs.items():
+            if par in self.params.__dict__:
+                setattr(self.params, par, v)
+        self.labels = pd.read_csv(LABELS_PATH)['display_name'].tolist()
+        self.label_subset = label_subset
+        self.top_n = top_n
+        if self.label_subset:
+            for l in self.label_subset:
+                if l not in self.labels:
+                    logging.warning('''Label {} does not exist. 
+                                     Dropping.'''.format(l))
+        super(AudiosetLabelExtractor, self).__init__()
+
+    def _extract(self, stim):
+        params = self.params
+        params.SAMPLE_RATE = stim.sampling_rate
+
+        if params.SAMPLE_RATE >= 2 * params.MEL_MAX_HZ:
+            if params.SAMPLE_RATE != 16000:
+                logging.warning(
+                    'The sampling rate of the stimulus is '
+                    f'{params.SAMPLE_RATE}Hz. '
+                    'YAMNet was trained on audio sampled at 16000Hz. ' 
+                    'This should not impact predictions, but you can resample ' 
+                    'the input using AudioResamplingFilter for full conformity ' 
+                    'to training.')
+            if params.MEL_MIN_HZ != 125 or params.MEL_MAX_HZ != 7500:
+                logging.warning(
+                    'Custom values for MEL_MIN_HZ and MEL_MAX_HZ '
+                    'were passed. Changing these defaults might affect ' 
+                    'model performance.')
+        else:
+            raise ValueError(
+                f'The sampling rate of your stimulus ({params.SAMPLE_RATE}Hz)'
+                ' must be at least twice the value of MEL_MAX_HZ '
+                f'({params.MEL_MAX_HZ}Hz). ' 
+                'Upsample your audio stimulus (recommended) or pass a lower '
+                'value of MEL_MAX_HZ when initializing this extractor.')
+
+        labels = self.labels
+        model = yamnet.yamnet_frames_model(params)
+        model.load_weights(self.weights_path)
+        preds, _ = model.predict_on_batch(np.reshape(stim.data, [1,-1]))
+        preds = preds.numpy()
+
+        if self.label_subset:
+            for l in self.label_subset:
+                print(l)
+                print(np.where(labels == l))
+            label_subset_idx = [idx for idx, lab in enumerate(labels) 
```

Comment:
The logic of which labels are kept and how the indexes are handled gets a bit confusing.

I would try to handle as much of this upfront in the `__init__` function. I would make `top_n` and `labels_subset` mutually exclusive arguments, such that specifying both raises an error. 

Then, compute the idxs up front for the subset_labels, and for `top_n` I suppose you have to compute that in `_extract` as it depends on the results.

Reply:
`__init__` now raises an error if both `label_subset` and `top_n` are not `None`.
In `__init__`, a `self.label_idx` variable is also computed which has the indices of the labels in `label_subset` (if this exists), else it has all label indices.
This is used to subset the set of predictions and labels in `_extract` without using conditional statements checking whether `label_subset` argument has been passed.

If `top_n` is defined, outputs and labels are further subsetted in `_extract`.

Note that no matter whether `top_n` is defined or not, outputs are always returned so to have labels with higher predictions on avg appear earlier (left in wide format; lines 619-622).